### PR TITLE
WIP rr graph lib

### PR DESCRIPTION
**Pull request by mcmasterg**
This PR is towards completing a baseline python rr_graph library to make custom rr_graphs beyond what arch.xml can do.

Current test suite is something like:
python -m lib.rr_graph.channel: passes
python -m lib.rr_graph.graph: passes
python utils/testarch_graph.py tests/build/testarch/2x4/wire.rr_graph.xml: fails due to missing channels (being investigated)

After stabilizing above to create the expected data structure, a relatively small amount of work needs to be done to add to_xml() functions to assorted XML elements. For the final library there might be a little work related to open VPR PR on mapping pin names to pins in the rr_graph pins (which is ambiguous in current format).

### CODE REVIEW

File: utils/lib/rr_graph/channel.py [hunk 1]
```diff
@@ -14,100 +19,120 @@
 class ChannelNotStraight(TypeError):
     pass
 
+# TODO: clean up type
+# real graphs have length 1 tracks, which confuses auto detection
+_Track = namedtuple("Track", ("start", "end", "idx", "type_hint"))
```
Comment (mithro) [inline utils/lib/rr_graph/channel.py]: I don't understand why `type_hint` is needed? Is it for "zero length" tracks?

File: utils/lib/rr_graph/channel.py [hunk 2]
```diff
@@ -243,318 +268,516 @@ def __str__(self):
 
 
 # Nice short alias..
-C = Channel
+T = Track
 
 
 class ChannelGrid(dict):
+    '''
+    Functionality:
```
Comment (mithro) [inline utils/lib/rr_graph/channel.py]: FYI - Docstring comments should be [reStructuredText](http://docutils.sourceforge.net/docs/ref/rst/introduction.html). 

If you write mostly markdown style it'll work okay.

File: utils/lib/rr_graph/channel.py [hunk 3]
```diff
@@ -243,318 +268,516 @@ def __str__(self):
...
+    -Manages single type of track (either chanx or chany)
+    -Channel width along grid
+    -Manages track allocation within channels
+    -A track allocator
+
+    dict is indexed by Pos() objects
+    This returns a list indicaitng all the tracks at that position
```
Comment (mithro) [inline utils/lib/rr_graph/channel.py]: speling

File: utils/lib/rr_graph/channel.py [hunk 4]
```diff
@@ -243,318 +268,516 @@ def __str__(self):
...
-            for y in range(0, self.y):
+        for x in range(0, self.width):
+            for y in range(0, self.height):
                 self[Pos(x,y)] = []
 
     @property
-    def x(self):
-        return self.size.x
+    def width(self):
+        """Grid width
```
Comment (mithro) [inline utils/lib/rr_graph/channel.py]: Maybe something like "Number of channels (as opposed to the width a an individual channel?)"

File: utils/lib/rr_graph/channel.py [hunk 5]
```diff
@@ -243,318 +268,516 @@ def __str__(self):
...
+        [T(B,0), T(T,1), T(E,2), T(G,3)]
         >>> g[(6,5)]
-        [C(B,0), C(C,1), None, C(G,3)]
+        [T(B,0), T(T,1), None, T(G,3)]
         >>> g[(7,5)]
         [None, None, None, None]
         """
-        assert ch.idx == None
+        assert t.idx == None
+        force_idx = idx
```
Comment (mithro) [inline utils/lib/rr_graph/channel.py]: I would probably have a `set_track` and a `add_track` -- only the set_track allows an idx value to be set?

File: utils/lib/rr_graph/channel.py [hunk 6]
```diff
@@ -243,318 +268,516 @@ def __str__(self):
...
+            p[t.idx] = t
+        return t
 
     def pretty_print(self):
         """
-        If type == Channel.Type.X
+        If type == Track.Type.X
 
-          A--AC-C
+          A--AC-T
```
Comment (mithro) [inline utils/lib/rr_graph/channel.py]: Search and replace fail....
`A--AC-C`

File: utils/lib/rr_graph/channel.py [hunk 7]
```diff
@@ -243,318 +268,516 @@ def __str__(self):
...
+        return s
 
 
 if __name__ == "__main__":
     import doctest
+    print('doctest: begin')
     doctest.testmod()
+    print('doctest: end')
+
+    if 1:
```
Comment (mithro) [inline utils/lib/rr_graph/channel.py]: Maybe these are better as unit tests?

File: utils/lib/rr_graph/graph.py [hunk 8]
```diff
@@ -1,4 +1,44 @@
...
+    xml: can import but doesn't keep track of a node
+class Block(MostlyReadOnly):
+    for <grid_loc>
+class BlockGrid:
+    Was: BlockGraph
+    xml: nothing, handled by intneral Block objects though
+class GraphIdsMap:
+    holds pins + edges
+    xml: updated as pins are added
+        inconsistent with the rest of the project
```
Comment (mithro) [inline utils/lib/rr_graph/graph.py]: FYI -- The reason for the inconsistency is that the other parts are not designed to allow modification/changing the XML.

File: utils/lib/rr_graph/graph.py [hunk 9]
```diff
@@ -783,59 +864,25 @@ def block_types_for(self, col=None, row=None):
             ss.append(self.block_grid[pos].block_type)
         return ss
 
-    def __getitem__(self, pos):
-        return self.block_grid[pos]
-
-
-def simple_test_graph():
```
Comment (mithro) [inline utils/lib/rr_graph/graph.py]: Where did this go?

File: utils/lib/rr_graph/graph.py [hunk 10]
```diff
@@ -1312,13 +1480,302 @@ def add_channel(self, globalname, start, end, segtype, _chantype=None):
...
+        self.ids.add_edge(int(xtrack_node.get("id")), int(ytrack_node.get("id")), int(switch.get("id")))
+        self.ids.add_edge(int(ytrack_node.get("id")), int(xtrack_node.get("id")), int(switch.get("id")))
+
+    def create_xy_track(self, start, end, idx=None, type=None):
+        '''Create track object and corresponding nodes'''
+        track = self.channels.create_xy_track(start, end, idx=idx, type=type)
+        track_node = self.ids.add_node_for_track(track)
+        return track, track_node
+
+    def index_node_objects(self):
```
Comment (mithro) [inline utils/lib/rr_graph/graph.py]: I don't get the point of this function?

File: utils/lib/rr_graph/channel.py (hunk 2 above)
Comment (mcmasterg) [reply utils/lib/rr_graph/channel.py]: I'm not paying any attention to docstring stuff right now. I'll sort that out later

**Comment by mithro** [comment]
The pull request at https://github.com/SymbiFlow/symbiflow-arch-defs/pull/115 has replaced this one.

**Pull request closed without merging**